File: recursion/lists/applied/merge_two_lists.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def mergeTwoLists(l1: ListNode, l2: ListNode): ListNode = {
        (l1, l2) match {
            // base case
            case (null, null) => null
            // different lengths etc.
            case (l1, null) => l1
            case (null, l2) => l2
            // recursive case
            case (l1, l2) if l1.x < l2.x => ListNode(l1.x, mergeTwoLists(l1.next, l2))
            case (l1, l2) => ListNode(l2.x, mergeTwoLists(l1, l2.next))
        }
    }
}

File: data_structures/in_place/merge_two_lists.go
package data_structures

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	// choice: 1. for loop 2. recursion
	// choice: 1. in-place 2. new result
	// choice: 1. build from tail : more natural recursion but O(2n)
	//         2. build from head harder but single pass splicing in l2 values
	list1 = &ListNode{0, list1} // add to head of list so can use .Next test
	result := list1             // need head of list
	for list1.Next != nil {
		if list2 != nil {
			if list1.Next.Val > list2.Val {
				list1.Next = &ListNode{list2.Val, list1.Next} // splice in list2 val between list1 and l1.Next
				list2 = list2.Next                            // move onto next l2
			}
		}
		list1 = list1.Next
	}
	list1.Next = list2 // if list2 has any remaining elements add them on to the end
	return result.Next
}
